Benqi-Agents/Large-Grant/src/agent.ts

```diff
@@ -0,0 +1,83 @@
+import { BigNumber, Contract } from "ethers";
+import { Interface } from "ethers/lib/utils";
+import { JsonRpcProvider } from "@ethersproject/providers/lib/json-rpc-provider";
+import { Finding, getEthersProvider, HandleTransaction, LogDescription, TransactionEvent } from "forta-agent";
+import CONFIG from "./agent.config";
+import { createFinding } from "./finding";
+import { AgentConfig, QI_BALANCE_ABI, QI_GRANTED_ABI, QI_TOTAL_SUPPLY, ThresholdMode } from "./utils";
+
+const QI_IFACE = new Interface([QI_BALANCE_ABI]);
+
+let comptrollerBalance = BigNumber.from(-1);
+
+const provideIsLarge = (agentConfig: AgentConfig): ((value: BigNumber) => boolean) => {
+  const bnThreshold = BigNumber.from(agentConfig.threshold);
+
+  switch (agentConfig.thresholdMode) {
+    case ThresholdMode.ABSOLUTE:
+      return (value: BigNumber): boolean => {
+        return value.gte(bnThreshold);
+      };
+    case ThresholdMode.PERCENTAGE_TOTAL_SUPPLY:
+      return (value: BigNumber): boolean => {
+        const totalSupply = BigNumber.from(QI_TOTAL_SUPPLY);
+        return value.gte(totalSupply.mul(bnThreshold).div(100));
+      };
+    case ThresholdMode.PERCENTAGE_COMPTROLLER_BALANCE:
+      return (value: BigNumber): boolean => {
+        return value.gte(comptrollerBalance.mul(bnThreshold).div(100));
+      };
+  }
+};
+
+const provideUpdateComptrollerBalance = (
+  agentConfig: AgentConfig,
+  provider?: JsonRpcProvider
+): ((block: number) => Promise<void>) => {
+  let balanceBlock = -1;
+
+  if (agentConfig.thresholdMode !== ThresholdMode.PERCENTAGE_COMPTROLLER_BALANCE) {
+    return async () => {};
+  } else {
+    return async (block: number) => {
+      if (block !== balanceBlock) {
+        balanceBlock = block;
+
+        const qiContract = new Contract(agentConfig.qiAddress, QI_IFACE, provider);
+        comptrollerBalance = await qiContract.balanceOf(agentConfig.comptrollerAddress, {
+          blockTag: block - 1,
+        });
```

Comment:
Instead of having that `comptrollerBalance` global variable, you can use the cache to save many values. then use it in provideUpdateComptrollerBalance  instead of ` if (block !== balanceBlock)` logic.